steely/local_steely.py

```diff
@@ -0,0 +1,101 @@
+import argparse
+from utils import scan_plugins_dir, load_plugin
+
+class FBChatMessageMock:
+    def __init__(self, message):
+        self.text = message
+
+
+class FBChatMock:
+    def __init__(self, prev_messages):
+        # List of messages, oldest to newest.
+        # This list will be updated with new messages so anything holding a 
+        # reference to it will be up to date.
+        self.prev_mes = prev_messages
+        # This is a seperate list for isolating the output of the plugin.
+        self.output_mes = []
+
+    def fetchThreadMessages(self, thread_id, limit):
+        return [FBChatMessageMock(m) for m in self.prev_mes[-limit:]][::-1]
+
+    def sendMessage(self, mess, thread_id, thread_type):
+        self.output_mes.append(mess)
+        self.prev_mes.append(mess)
+
+
+class SteelyREPL:
+    def __init__(self, plugins, prefix='.', author_id="test_author"):
+        self._plugins = {
+                prefix + p.COMMAND: p for p in plugins if p.COMMAND is not None
+        }
+        self._all_message_plugins = [p for p in plugins if p.COMMAND is None]
+        self._message_list = []
+        self._author_id = author_id
+
+    def _exec_plugin(self, plugin, message):
+        bot = FBChatMock(self._message_list)
+        plugin.main(bot, self._author_id, message,
+                    thread_id='test_thread', thread_type='test_thread_type')
+        return bot.output_mes
+
+    def _exec_message(self, message):
+        plugin_name, *arg = message.split(maxsplit=1)
+        arg_str = arg[0] if arg else ''
+        if plugin_name in self._plugins:
+            yield from self._exec_plugin(self._plugins[plugin_name], arg_str)
+            return
+
+        for plugin in self._all_message_plugins:
+            yield from self._exec_plugin(plugin, message)
+
+    def _run_noguard(self):
```

Comment:
This method ordering  would make more sense:
```
run
_run_noguard
_exec_message
_exec_plugin
```
(step-down)

Reply:
Hm, to me the opposite is true. Coming from C/C++ where functions are declared least dependant to most dependant it's unexpected to see them in a different order.

Reply:
Eh fair enough, although that's an old practice that became less common when forward-declarations were no longer required.

Reply:
I can't find much in terms of a "step up rule", but a lot of people (including me) advocate for the opposite.

Reply:
I've looked around, and there's no standard for method order in Python. I agree with Cian here, mostly for personal preference. However, it also seems reasonable to fall back to the C standard when there is no Python one.